pkg/probe/grpc/grpc.go

```diff
@@ -47,14 +47,16 @@ func New() Prober {
 
 // Probe executes a grpc call to check the liveness/readiness/startup of container.
 // Returns the Result status, command output, and errors if any.
-// Only return non-nil error when service is unavailable and/or not implementing the interface,
-// otherwise result status is failed,BUT err is nil
-func (p grpcProber) Probe(host, service string, port int, timeout time.Duration, opts ...grpc.DialOption) (probe.Result, string, error) {
+// Any failure is considered as a probe failure to mimic grpc_health_probe tool behavior.
+// err is always nil
+func (p grpcProber) Probe(host, service string, port int, timeout time.Duration) (probe.Result, string, error) {
 	v := version.Get()
 
-	md := metadata.New(map[string]string{
-		"User-Agent": fmt.Sprintf("kube-probe/%s.%s", v.Major, v.Minor),
-	})
+	opts := []grpc.DialOption{
+		grpc.WithUserAgent(fmt.Sprintf("kube-probe/%s.%s", v.Major, v.Minor)),
+		grpc.WithBlock(),
+		grpc.WithInsecure(), //credentials are currently not supported
```

Comment:
This comment seems misleading to me. `WithInsecure()` isn't a credentials option, this appears to completely disable transport security: https://github.com/grpc/grpc-go/blob/d542bfcee46d733f7bf8a5e870994379863da2d2/clientconn.go#L79-L82

The KEP says that gRPC probes will only support HTTP/2 over TLS, and that we will _not_ support unencrypted HTTP/2, but I believe this completely disables TLS: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2727-grpc-probe#proposal

I am not comfortable with disabling this by default without further input given that this hasn't been discussed. I can understand why it would be in the tests, but this appears to disable it globally, including in production.

/hold

Reply:
grpc only allows two modes - `WithInsecure()` or one of secure options. See https://pkg.go.dev/google.golang.org/grpc#WithInsecure. Since the KEP was not intended to provide any secure connection health checks mechanism, and this is in line with what http proves do with TLS, this change is in spirit of the KEP intended description. 

Also the expectation is that people will be migrating from https://github.com/grpc-ecosystem/grpc-health-probe/blob/b0cf9b4985ab201654da993ea90a7b459527fadb/main.go#L225-L252 which makes this the only viable option to be implemented in this KEP.

@bowei wdyt?

Reply:
> grpc only allows two modes - WithInsecure() or one of secure options. See https://pkg.go.dev/google.golang.org/grpc#WithInsecure.

Just to be clear, adding this flag here is the only change that MUST be merged to make feature usable. Without this flag, feature is not working - grpc client is not able to connect to server AT ALL

Reply:
I think we support HTTPS as an option for `httpGet` probes.  I don't know gRPC well - what are the secure options and why are they not viable?

I suspect if we went to alpha without them, but added secure for beta it might be OK - is that possible?

Reply:
I can understand why we wouldn't want mTLS but I'm not sure why regular TLS connections aren't supportable.

Just to quote the KEP from the [link above](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2727-grpc-probe#proposal),

> This will result in the use of gRPC (using HTTP/2 over TLS) to use the standard healthcheck service to determine the health of the container. As spec'd, the kubelet probe will not allow use of client certificates nor verify the certificate on the container. We do not support other protocols for the time being (unencrypted HTTP/2, QUIC).

I think if @bowei @dchen1107 @derekwaynecarr are okay with changing this, then there's no issue. I just wanted to note that this is the opposite of the approved spec (and I don't have the approver powers to update it).

Reply:
+1 - insecure || (unilateral) TLS seems OK to me, and I'd be OK with TLS coming after alpha if needed

Reply:
> I think we support HTTPS as an option for `httpGet` probes. I don't know gRPC well - what are the secure options and why are they not viable?

confirmed that we support HTTPS (but with insecure verification) for probes since 2015: https://github.com/kubernetes/kubernetes/pull/9965

Reply:
I think it’s fine to use insecure for alpha.  I was hoping this feature would offer support for secure by the time it’s enabled for beta.  As I understand it now, this makes the feature consistent with existing HTTPS probe support.